### ISSUE TITLE
Travis tests are broken

### ISSUE THREAD

**Issue by zachriggle** [Gallopsled/pwntools#1133]
Currently, the Travis CI tests for `dev` are erroring out.  I don't have any cycles to investigate this right now.

**Comment by zachriggle** [comment]
@Idolf any chance you can take a look?

**Comment by bruce30262** [comment]
@zachriggle I've take a look at the latest build of the dev branch ( [link](https://travis-ci.org/Gallopsled/pwntools/builds/363799792?utm_source=github_status&utm_medium=notification) ) and it seems that it's because travis-ci no longer support ubuntu 12.04, causing the build failed to install the addons (  e.g. `as` ). This will failed all the tests that required the assembler ( `runner.rst`, `elf.rst`.....)  

Right now I'm working on a fix. I've switch to the trusty build and add the `sudo: required` option in `.travis.yml`, and the amount of the fail tests has immediately down to only 2 failed tests, you can view the log [here](https://travis-ci.org/bruce30262/pwntools/jobs/383528684). One of them is `p.leak()`, which you've already mentioned in the [readme of travis's docker](https://github.com/Gallopsled/pwntools/tree/dev/travis/docker). Another one has the following error:  
```
File "shellcraft/mips.rst", line ?, in default
Failed example:
    print enhex(asm(shellcraft.mips.sh()))
Exception raised:
    Traceback (most recent call last):
      File "/opt/python/2.7.14/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        print enhex(asm(shellcraft.mips.sh()))
    TypeError: 'unicode' object is not callable
```
Looks like it's some error about encoding. I'm currently working on this test.

**Comment by zachriggle** [comment]
First, thanks for looking into this. I wish I had more time to give back to
the community to fix these issues, but it’s just not possible right now. My
CTF team qualified for DEFCON finals (alas, I did not participate in Quals)
so maybe I can poke some of them to do it.

If you add “sudo: true” to the config, it should fix the first error.

For the second one, I’d suggest getting a Trusty minimal server install VM
and run travis/install.sh and then make -C docs doctest.

See TESTING.md for more info.

**Comment by bruce30262** [comment]
@zachriggle  Got some good news here , the [build](https://travis-ci.org/bruce30262/pwntools/jobs/383558274) in travis-ci finally pass all the test cases :)))

----------------
The `shellcraft/mips` error got me confuse though.  
For testing `shellcraft.XXX.linux.sh()` ( XXX = aarch64, amd64....) , most of them are using the following test case:  
```
>>> p = run_assembly(shellcraft.XXX.linux.sh())
>>> p.sendline('echo Hello')
>>> p.recv()
'Hello\n'
``` 
However, `shellcraft.mips.linux.sh()` is using this one:  
```
>>> print enhex(asm(shellcraft.mips.sh()))
6269093c2f2f2935f4ffa9af7368093c6e2f2935f8ffa9affcffa0aff4ffbd272020a003fcffa0affcffbd27ffff0628fcffa6affcffbd232030a00373680934fcffa9affcffbd27ffff0528fcffa5affcffbd23fbff1924272820032028bd00fcffa5affcffbd232028a003ab0f02340c010101
```
🤔🤔🤔???  I wonder why we're verifying the hex string.....

Anyway after I changed it to:
```
>>> p = run_assembly(shellcraft.mips.linux.sh())
>>> p.sendline('echo Hello')
>>> p.recv()
'Hello\n'
```
It pass the test case: [travis log](https://travis-ci.org/bruce30262/pwntools/builds/383546316)

----------------------------
For the `p.leak()` error, after I add the `sudo:true` option, the build still failed in travis-ci:  
```
stack smashing detected**********************************************************************
File "tubes/processes.rst", line ?, in default
Failed example:
    p.leak(e.address, 4)
Exception raised:
    Traceback (most recent call last):
      File "/opt/python/2.7.14/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        p.leak(e.address, 4)
      File "/home/travis/build/bruce30262/pwntools/pwnlib/tubes/process.py", line 956, in leak
        return mem.read(count) or None
    IOError: [Errno 5] Input/output error
```
And so I started testing on my Ubuntu VM, and found that even if I'm root, I still can't read the content of `/bin/sh`'s `/proc/self/mem`. I'm guessing that the IO between the process and `/bin/sh`'s `/proc/self/mem` has been blocked for some unknown reason. Then I changed the test binary of process.leak() from `/bin/sh` to `/bin/bash`, and found that it work -- both on my VM and travis-ci  :))))  

-------------------------------

Now I would like to ask some questions before I create a PR:

1. Since both `sudo: required` and `sudo: true` work on the latest build, which one should we use ?  
2. Any thought on these fixes ?

**Comment by zachriggle** [comment]
For “sudo” just pick, I don’t have a preference. What do the Travis docs
suggest?

For the “run the code instead of checking its output”, sounds good to me!

Please create a pull request against the “stable” branch. You might need to
cherry-pick your fixes against that branch if you were working on “dev”.

**Comment by bruce30262** [comment]
PR has been created ( #1155 ), please check it out .

**Comment by zachriggle** [comment]
Fixed via #1155
